reward/staking_manager.go

```
@@ -195,6 +213,71 @@ func updateStakingInfo(blockNum uint64) (*StakingInfo, error) {
+	if err != nil {
+		return nil, fmt.Errorf("failed to call AddressBook contract. root err: %s", err)
+	}
+
+	if len(types) == 0 && len(addrs) == 0 {
+		// This is an expected behavior when the addressBook contract is not activated yet.
+		logger.Info("The addressBook is not yet activated. Use empty stakingInfo", "reason", err)
+		return newEmptyStakingInfo(blockNum), nil
+	}
+
+	if len(types) != len(addrs) {
+		logger.Error("Fail while parsing a result from the addressBook. Use empty staking info", "err", fmt.Errorf("length of type list and address list differ. len(type)=%d, len(addrs)=%d", len(types), len(addrs)))
+		return newEmptyStakingInfo(blockNum), nil
+	}
+
+	var (
+		nodeIds      = []common.Address{}
+		stakingAddrs = []common.Address{}
+		rewardAddrs  = []common.Address{}
+		pocAddr      = common.Address{}
+		kirAddr      = common.Address{}
+	)
+
+	// Parse and construct node information
+	for i, addrType := range types {
+		switch addrType {
+		case addressTypeNodeID:
+			nodeIds = append(nodeIds, addrs[i])
+		case addressTypeStakingAddr:
+			stakingAddrs = append(stakingAddrs, addrs[i])
+		case addressTypeRewardAddr:
+			rewardAddrs = append(rewardAddrs, addrs[i])
+		case addressTypePoCAddr:
+			pocAddr = addrs[i]
+		case addressTypeKIRAddr:
+			kirAddr = addrs[i]
+		default:
+			logger.Error("Fail while parsing a result from the addressBook. Use empty staking info", "err", fmt.Errorf("invalid type from AddressBook: %d", addrType))
+			return newEmptyStakingInfo(blockNum), nil
+		}
+	}
+
+	// validate parsed node information
+	if len(nodeIds) != len(stakingAddrs) ||
+		len(nodeIds) != len(rewardAddrs) ||
+		common.EmptyAddress(pocAddr) ||
+		common.EmptyAddress(kirAddr) {
+		// This is an expected behavior when the addressBook contract is not activated yet.
+		logger.Info("The addressBook is not yet activated. Use empty stakingInfo", "reason", err)
```

No `err` exist for this context. And, this code is not reachable if AddressBook is not activated. Don't we need to return an error for this case?

---

https://github.com/klaytn/klaytn/blob/2dc0b6578b6def6e80238351397170e92d747b0d/reward/addressbook_connector.go#L155-L160

---

In the previous code, an error log was printed and `newEmptyStakingInfo(blockNum)` was returned

---

@blukat29 handled this in #1882